python/ray/scripts/scripts.py

```diff
@@ -1862,6 +1866,233 @@ def local_dump(
+)
+@click.option(
+    "--actor-id",
+    "-a",
+    required=False,
+    type=str,
+    default=None,
+    help="Retrieves the logs corresponding to this ActorID.",
+)
+@click.option(
+    "--task-id",
+    "-t",
+    required=False,
+    type=str,
+    default=None,
+    help="Retrieves the logs corresponding to this TaskID.",
+)
+@click.option(
+    "--watch",
+    "-w",
+    required=False,
+    type=bool,
+    is_flag=True,
+    help="Streams the log file as it is updated instead of just tailing.",
+)
+@click.option(
+    "--lines",
+    "-l",
+    required=False,
+    type=int,
+    default=None,
+    help="Number of lines to tail from log. -1 indicates fetching the whole file.",
+)
+def logs(
+    filters,
+    filename: str,
+    node_ip: str,
+    node_id: str,
+    pid: str,
+    actor_id: str,
+    task_id: str,
+    watch: bool,
+    lines: int,
+):
+    """
+    View logs in the ray cluster.
+
+    FILTERS: Filename substrings (filename, component, file extension, id) to filter
+    the logs by.
+
+    Example usage:
+
+    ray logs -a <actor-id>               # Display worker stdout log by actor-id
+
+    ray logs -ip 198.0.0.1 -pid 98712    # Display worker stdout log by ip & pid
+
+    ray logs -n <node-id> -f raylet.out  # Display log by filename & node-id
+
+    ray logs                             # Display list of logs by category
+
+    ray logs worker .out <worker-id>     # Filter logs by filename substring
+
+    ray logs --endpoint=198.0.0.1:8265   # Retrieves logs from a remote cluster
+
+    """
+
+    try:
+        found_many = False
+
+        if task_id is not None:
+            raise ValueError("--task-id is not yet supported")
+
+        match_node = node_ip is not None or node_id is not None
+        match_file = filename is not None or pid is not None
+        match_actor = actor_id is not None
+
+        match_unique = (match_node and match_file) or match_actor
+
+        if match_file and not match_node:
+            identifier = f"filename: {filename}" if filename else f"pid: {pid}"
+            raise ValueError(
+                f"Unique logfile identifier '{identifier}' needs to be "
+                "accompanied with a node identifier (--node-id or --node-ip)."
+            )
+
+        if not match_unique:
+            # Try to match a single log file.
+            # If we find more than one match, we output the index.
+            filters = ",".join(filters) + (
+                f",{filename}" if filename is not None else ""
+            )
+            api_endpoint, logs_dict = list_logs(node_id, node_ip, filters)
+            # to_dedup = ["gcs_logs", "dashboard", "autoscaler", "autoscaler_monitor"] ?
+            if len(logs_dict) == 0:
+                raise Exception("Could not find node.")
+            if filename is None:
+                for node_id, logs in logs_dict.items():
+                    log = None
+                    for log_list in logs.values():
```

Comment:
What about

```python3
log_matches = list (filter(lambda log_files: len(log_files) > 0, log_list_per_node.values()))
if not log_matches:
    raise ...
if log_matches == 1:
    matched_file = log_matches[0]
else:
    matched_file = None

if not matched_file:
    # List files
else:
    # Print
```

Reply:
Hmm maybe we need to flatten `log_matches ` one more time

Reply:
Used sum(nested, [])